### ISSUE TITLE
DataCollatorForLanguageModeling always uses pad_side='right'

### ISSUE THREAD

**Issue by kimihailv** [huggingface/trl#4497]
### Reproduction

```python
from trl.trainer.sft_trainer import DataCollatorForLanguageModeling
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("google/gemma-3-4b-it")

collator = DataCollatorForLanguageModeling(
    tokenizer.pad_token_id,
)

print(tokenizer.padding_side)

batch = tokenizer.apply_chat_template(
    [
        [
            {"role": "user", "content": "Hello, how are you?"},
            {"role": "assistant", "content": "I'm fine, thank you!"}
        ],
        [{"role": "user", "content": "Hello, how are you?"}]
    ],
    tokenize=True,
)

batch = collator([{"input_ids": batch[0]}, {"input_ids": batch[1]}])
print(batch)
```

outputs:

```
left
{'input_ids': tensor([[     2,    105,   2364,    107,   9259, 236764,   1217,    659,    611,
          236881,    106,    107,    105,   4368,    107, 236777, 236789, 236757,
            5851, 236764,   7806,    611, 236888,    106,    107],
         [     2,    105,   2364,    107,   9259, 236764,   1217,    659,    611,
          236881,    106,    107,      0,      0,      0,      0,      0,      0,
               0,      0,      0,      0,      0,      0,      0]]),
 'labels': tensor([[     2,    105,   2364,    107,   9259, 236764,   1217,    659,    611,
          236881,    106,    107,    105,   4368,    107, 236777, 236789, 236757,
            5851, 236764,   7806,    611, 236888,    106,    107],
         [     2,    105,   2364,    107,   9259, 236764,   1217,    659,    611,
          236881,    106,    107,   -100,   -100,   -100,   -100,   -100,   -100,
            -100,   -100,   -100,   -100,   -100,   -100,   -100]]),
 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
          1],
         [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0]])}
```

### System Info

- Platform: Linux-5.15.0-140-generic-x86_64-with-glibc2.35
- Python version: 3.11.13
- TRL version: 0.25.0
- PyTorch version: 2.8.0
- accelerator(s): NVIDIA H100 80GB HBM3
- Transformers version: 4.56.2
- Accelerate version: 1.10.1
- Accelerate config: not found
- Datasets version: 4.1.1
- HF Hub version: 0.35.0
- bitsandbytes version: 0.47.0
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.17.1
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by kimihailv** [comment]
The reason: https://github.com/huggingface/trl/blob/32febb491b386881fe75137e9990fb0f1b5cae2c/trl/trainer/sft_trainer.py#L195-L221

**Comment by pramodith** [comment]
@kimihailv I believe that this is the intended behavior for SFT. In SFT you never really generate tokens, all you do is run a single forward pass over the entire input, obtain the prob scores for each of those tokens and compute cross entropy.

You need to pad to the left when you need to generate text.

**Comment by kimihailv** [comment]
@pramodith To infer with left padding you need to train the model working with such inputs. Yes, attention masks force zero attention to padding tokens, but nevertheless

**Comment by behroozazarkhalili** [comment]
I looked into this, and I believe @pramodith is right: right-padding is the intended and correct behavior for the SFT loss collator. Here is the check.

## What the collator does

In the non-`padding_free` path, `DataCollatorForLanguageModeling` emits `input_ids`, `labels` (padding masked to `-100`), and an `attention_mask` — it does **not** emit `position_ids`. The model derives positions from the attention mask, so padding tokens do not participate in the loss (masked labels) or in attention (zeroed mask).

## Does padding side change SFT results?

I padded the same short sequence right vs left and compared the real-token logits to the **unpadded** sequence (tiny model, single forward):

| padding | real-token logits vs unpadded reference |
|---|---|
| right | max abs diff = `0.0` |
| left | max abs diff = `2.4e-4` (fp32 ordering noise) |

Both reproduce the unpadded computation. The small loss/grad differences you can observe between the two are floating-point ordering artifacts, not a difference in what the model learns. Since SFT is a single forward pass + cross-entropy over the (label-masked) input and never generates, the padding side does not affect training.

## The inference concern

Your underlying point — that you'd want training and inference padding to match — is reasonable, but it is a **generation-time** concern, not an SFT-loss one. Left-padding matters when you `generate` (so the last real token is at the sequence end); TRL's eval/generation paths already use left-padding for that. The SFT training collator computing the LM loss is a separate step where right-padding is correct.

## Suggestion

I think this can be closed as working-as-intended: right-padding in the SFT collator does not affect training results, and the model handles positions via the attention mask. Honoring `tokenizer.padding_side` here would add a configuration knob with no effect on the loss.

**Comment by albertvillanova** [comment]
Hey @kimihailv, sorry for the slow reply — we're catching up on the backlog. 

This is intended (as @behroozazarkhalili noted as well): the SFT collator always right-pads, regardless of the tokenizer's `padding_side` (see https://github.com/huggingface/trl/blob/0ed4ac7fa03bf388a94566f1b1f71e71445ca6e7/trl/trainer/sft_trainer.py#L524-L527). As @pramodith noted, SFT runs a single forward pass over the full sequence rather than generating, so the `attention_mask` already prevents pad tokens from contributing to the loss. Right-padding is the safe default for training because the real tokens stay left-aligned, keeping the default `arange` position IDs correct; left-padding would silently shift them unless position IDs were recomputed. Training padding side is also independent of inference: batched generation uses left-padding and Transformers handles its position IDs separately, so you don't need to train left-padded to infer left-padded. 

Closing as not-a-bug, but happy to reopen if you have a concrete case where right-padding changes the SFT loss.
